pkg/resource/deploy/step_generator.go

```diff
@@ -1870,40 +1998,68 @@ func (sg *stepGenerator) providerChanged(urn resource.URN, old, new *resource.St
-	newInputs resource.PropertyMap, prov plugin.Provider, allowUnknowns bool,
-	ignoreChanges []string,
-) (plugin.DiffResult, error) {
+func (sg *stepGenerator) diff(
+	event RegisterResourceEvent,
+	goal *resource.Goal, autonaming *plugin.AutonamingOptions, randomSeed []byte,
+	urn resource.URN, old, new *resource.State, oldInputs,
+	newInputs resource.PropertyMap, prov plugin.Provider,
+) (plugin.DiffResult, *promise.CompletionSource[plugin.DiffResult], error) {
 	// If this resource is marked for replacement, just return a "replace" diff that blames the id.
 	if sg.isTargetedReplace(urn) {
-		return plugin.DiffResult{Changes: plugin.DiffSome, ReplaceKeys: []resource.PropertyKey{"id"}}, nil
+		return plugin.DiffResult{Changes: plugin.DiffSome, ReplaceKeys: []resource.PropertyKey{"id"}}, nil, nil
 	}
 
 	// Before diffing the resource, diff the provider field. If the provider field changes, we may or may
 	// not need to replace the resource.
 	providerChanged, err := sg.providerChanged(urn, old, new)
 	if err != nil {
-		return plugin.DiffResult{}, err
+		return plugin.DiffResult{}, nil, err
 	} else if providerChanged {
-		return plugin.DiffResult{Changes: plugin.DiffSome, ReplaceKeys: []resource.PropertyKey{"provider"}}, nil
+		return plugin.DiffResult{Changes: plugin.DiffSome, ReplaceKeys: []resource.PropertyKey{"provider"}}, nil, nil
 	}
 
 	// Apply legacy diffing behavior if requested. In this mode, if the provider-calculated inputs for a resource did
 	// not change, then the resource is considered to have no diff between its desired and actual state.
 	if sg.deployment.opts.UseLegacyDiff && oldInputs.DeepEquals(newInputs) {
-		return plugin.DiffResult{Changes: plugin.DiffNone}, nil
+		return plugin.DiffResult{Changes: plugin.DiffNone}, nil, nil
 	}
 
 	// If there is no provider for this resource (which should only happen for component resources), simply return a
 	// "diffs exist" result.
 	if prov == nil {
 		if oldInputs.DeepEquals(newInputs) {
-			return plugin.DiffResult{Changes: plugin.DiffNone}, nil
-		}
-		return plugin.DiffResult{Changes: plugin.DiffSome}, nil
-	}
-
-	return diffResource(urn, old.ID, oldInputs, oldOutputs, newInputs, prov, allowUnknowns, ignoreChanges)
+			return plugin.DiffResult{Changes: plugin.DiffNone}, nil, nil
+		}
+		return plugin.DiffResult{Changes: plugin.DiffSome}, nil, nil
+	}
+
+	if !sg.deployment.opts.ParallelDiff {
+		// If parallel diff isn't enabled just do the diff directly.
+		diff, err := diffResource(
+			urn, old.ID, oldInputs, old.Outputs, newInputs, prov, sg.deployment.opts.DryRun, goal.IgnoreChanges)
+		return diff, nil, err
+	}
+
+	// Else setup a promise for it so our caller will yield a DiffStep.
+	pcs := &promise.CompletionSource[plugin.DiffResult]{}
+	go func() {
+		// if promise had an "ContinueWith" we'd use it here, but a goroutine blocked on Result and then
+		// posting to a channel is very cheap.
```

Comment:
It's not clear to me what a `ContinueWith` would do/offer here, so I'm kinda on the fence about this comment -- either we flesh it out in the sense that it's a "hey if you're worried about the code below/think it's expensive etc., don't be", or we remove it. But maybe just me.